components/encryption/src/manager/mod.rs

```diff
@@ -98,75 +115,78 @@ impl Dicts {
         let file = EncryptedFile::new(&self.base, FILE_DICT_NAME);
-        let file_bytes = self.file_dict.write_to_bytes()?;
+        let file_bytes = file_dict.write_to_bytes()?;
         // File dict is saved in plaintext.
         file.write(&file_bytes, &PlaintextBackend::default())?;
 
         ENCRYPTION_FILE_SIZE_GAUGE
             .with_label_values(&["file_dictionary"])
             .set(file_bytes.len() as _);
-        ENCRYPTION_FILE_NUM_GAUGE.set(self.file_dict.files.len() as _);
+        ENCRYPTION_FILE_NUM_GAUGE.set(file_dict.files.len() as _);
 
         Ok(())
     }
 
-    fn get_key(&self, key_id: u64) -> Option<&DataKey> {
-        self.key_dict.keys.get(&key_id)
-    }
-
-    fn current_data_key(&self) -> (u64, &DataKey) {
-        (
-            self.key_dict.current_key_id,
-            self.key_dict
-                .keys
-                .get(&self.key_dict.current_key_id)
-                .unwrap_or_else(|| {
-                    panic!(
-                        "current data key not found! Number of keys {}",
-                        self.key_dict.keys.len()
-                    );
-                }),
-        )
-    }
-
-    fn get_file(&mut self, fname: &str) -> &FileInfo {
-        if self.file_dict.files.get(fname).is_none() {
+    fn get_file(&self, fname: &str) -> FileInfo {
+        let hash = get_index(fname);
+        let dict = self.sharded_file_dict[hash].lock().unwrap();
+        if dict.files.get(fname).is_none() {
+            drop(dict);
             // Return Plaintext if file not found
             let mut file = FileInfo::default();
             file.method = compat(EncryptionMethod::Plaintext);
-            self.file_dict.files.insert(fname.to_owned(), file);
+            let mut file_dict = self.file_dict.lock().unwrap();
+            file_dict.files.insert(fname.to_owned(), file.clone());
+            self.sharded_file_dict[get_index(fname)]
+                .lock()
+                .unwrap()
+                .files
+                .insert(fname.to_owned(), file.clone());
+            file
+        } else {
+            dict.files.get(fname).cloned().unwrap()
         }
-        self.file_dict.files.get(fname).unwrap()
     }
 
-    fn new_file(&mut self, fname: &str, method: EncryptionMethod) -> Result<&FileInfo> {
+    fn new_file(
+        &self,
+        fname: &str,
+        method: EncryptionMethod,
+        file_dict: &mut FileDictionary,
+        key_dict: &mut KeyDictionary,
+    ) -> Result<FileInfo> {
         let iv = Iv::new_ctr();
         let mut file = FileInfo::default();
         file.iv = iv.as_slice().to_vec();
-        file.key_id = self.key_dict.current_key_id;
+        file.key_id = key_dict.current_key_id;
         file.method = compat(method);
-        self.file_dict.files.insert(fname.to_owned(), file);
-        self.save_file_dict()?;
+        file_dict.files.insert(fname.to_owned(), file.clone());
```

Comment:
If the method is plaintext, update file dict is unnecessary?

Reply:
removed